File: NeoVM.Interop/NeoVM.Interop/Types/StackItems/InteropStackItem.cs
using NeoVM.Interop.Enums;
using NeoVM.Interop.Helpers;
using NeoVM.Interop.Interfaces;
using System;
using System.Runtime.InteropServices;

namespace NeoVM.Interop.Types.StackItems
{
    public class InteropStackItem : IStackItem<object>, IEquatable<InteropStackItem>
    {
        IntPtr ObjectPointer;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="engine">Engine</param>
        /// <param name="data">Data</param>
        internal InteropStackItem(ExecutionEngine engine, object data) : base(engine, data, EStackItemType.Interop)
        {
            CreateNativeItem();
        }
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="engine">Engine</param>
        /// <param name="handle">Handle</param>
        /// <param name="data">Data</param>
        /// <param name="ptr">Object pointer</param>
        internal InteropStackItem(ExecutionEngine engine, IntPtr handle, object data, IntPtr ptr) :
            base(engine, data, EStackItemType.Interop, handle)
        {
            ObjectPointer = ptr;
        }

        protected override void Free()
        {
            base.Free();

            if (ObjectPointer != IntPtr.Zero)
            {
                Marshal.Release(ObjectPointer);
                ObjectPointer = IntPtr.Zero;
            }
        }

        public bool Equals(InteropStackItem other)
        {
            return other != null && other.Value.Equals(Value);
        }

        public override bool Equals(IStackItem other)
        {
            if (other is InteropStackItem b)
                return b.Value.Equals(Value);

            return false;
        }

        protected override byte[] GetNativeByteArray()
        {
            IntPtr ptr = Marshal.GetIUnknownForObject(Value);
            return BitHelper.GetBytes(ptr.ToInt64());
        }
    }
}

File: src/Neo.HyperVM/InteropStackItem.cpp
#include "InteropStackItem.h"

bool InteropStackItem::Equals(IStackItem* it)
{
	if (it == this) return true;

	// Different type

	if (it->Type != EStackItemType::Interop)
		return false;

	auto ii = (InteropStackItem*)it;

	// Check content of the raw data

	if (ii->_payloadLength != this->_payloadLength)
		return false;

	for (int32 x = this->_payloadLength - 1; x >= 0; x--)
	{
		if (ii->_payload[x] != this->_payload[x])
			return false;
	}

	return true;
}